python/ray/data/_internal/compute.py

```diff
@@ -359,8 +394,50 @@ def map_block_nosplit(
                     block_indices[ref] = len(blocks_in)
                     tasks_in_flight[worker] += 1
 
+                # Try to scale down if there's no more work to do and pending/idle
+                # workers.
+                if not blocks_in:
+                    should_update = False
+                    if tasks_in_flight[worker] == 0:
+                        should_update = True
+                        # No more work to do and worker is idle, so we terminate the
+                        # worker.
+                        ready_workers.remove(worker)
+                        workers.remove(worker)
+                        # Explicitly terminate the actor to expedite cleanup.
+                        ray.kill(worker)
+                    # Terminate all pending workers, since we now know that they will
+                    # have no work to do. We try to eagerly terminate these pending
+                    # workers in order to:
+                    #  1. Free up resources they would otherwise acquire.
+                    #  2. Decrease Ray's scheduling load.
+                    #  3. Account for resource-bounded case where the actors will never
+                    #  be scheduled, in which case the actor creation will always be
+                    #  pending util the pool is destroyed.
+                    workers_to_terminate = workers - ready_workers
+                    if workers_to_terminate:
+                        # NOTE: This will run at most once in a pool's lifetime.
+                        should_update = True
+                        workers_to_ready_tasks = {}
+                        for task, worker in tasks.items():
+                            if worker in workers_to_terminate:
+                                # Worker is pending and should therefore only have a
+                                # single "ready" task in the task map.
+                                assert worker not in workers_to_ready_tasks
+                                workers_to_ready_tasks[worker] = task
+                        for worker_ in workers_to_terminate:
+                            workers.remove(worker_)
+                            del tasks[workers_to_ready_tasks[worker_]]
+                            # Explicitly terminate the actor to expedite cleanup.
+                            ray.kill(worker_)
```

Comment:
Without adding this eager pending worker termination, when the actor pool is resource-bound (not queue-bound), it scales up to 125% of the available resources, with 25% of the actors pending, and those pending actors then hang out for the rest of the run.
```
... (scale up to 32 bound)
Map Progress (32 actors 5 pending):   ...
...
Map Progress (32 actors 5 pending):   ...
... (pending actors continue to hang around)
```
Eagerly terminating these pending actors provided much better progress bar UX.

Reply:
We use `ray.kill()` here rather than graceful termination via `worker_.__ray_terminate__.remote()` since the former will cancel the actor creation task (lease request) while the latter would not, and would only run after the actor is scheduled, which defeats the purpose of this pending actor termination.